fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -27,197 +27,495 @@
+		auto const writeDesiredTps = getDesiredTps(tag, OpType::WRITE, averageTransactionWriteCost);
+		if (readDesiredTps.present() && writeDesiredTps.present()) {
+			return std::min(readDesiredTps.get(), writeDesiredTps.get());
+		} else if (readDesiredTps.present()) {
+			return readDesiredTps;
+		} else {
+			return writeDesiredTps;
+		}
+	}
+
+	Optional<double> getReservedTps(TransactionTag tag, OpType opType, double averageTransactionCost) const {
+		auto const reservedCost = getQuota(tag, opType, LimitType::RESERVED);
+		if (!reservedCost.present()) {
+			return {};
+		} else {
+			return reservedCost.get() / averageTransactionCost;
+		}
+	}
+
+	Optional<double> getReservedTps(TransactionTag tag,
+	                                double averageTransactionReadCost,
+	                                double averageTransactionWriteCost) const {
+		auto const readReservedTps = getReservedTps(tag, OpType::READ, averageTransactionReadCost);
+		auto const writeReservedTps = getReservedTps(tag, OpType::WRITE, averageTransactionWriteCost);
+		if (readReservedTps.present() && writeReservedTps.present()) {
+			return std::max(readReservedTps.get(), writeReservedTps.get());
+		} else if (readReservedTps.present()) {
+			return readReservedTps;
+		} else {
+			return writeReservedTps;
+		}
+	}
+
+	void removeUnseenQuotas(std::unordered_set<TransactionTag> const& tagsWithQuota) {
+		for (auto& [tag, stats] : tagStatistics) {
+			if (!tagsWithQuota.count(tag)) {
+				stats.clearQuota();
 			}
-			wait(delay(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_TRACE_INTERVAL));
 		}
 	}
 
 	ACTOR static Future<Void> monitorThrottlingChanges(GlobalTagThrottlerImpl* self) {
+		state std::unordered_set<TransactionTag> tagsWithQuota;
+
 		loop {
 			state ReadYourWritesTransaction tr(self->db);
-
 			loop {
-				// TODO: Clean up quotas that have been removed
 				try {
 					tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
 					tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
 
+					tagsWithQuota.clear();
 					state RangeResult currentQuotas = wait(tr.getRange(tagQuotaKeys, CLIENT_KNOBS->TOO_MANY));
-					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas").detail("Size", currentQuotas.size());
+					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas", self->id).detail("Size", currentQuotas.size());
 					for (auto const kv : currentQuotas) {
 						auto const tag = kv.key.removePrefix(tagQuotaPrefix);
 						auto const quota = ThrottleApi::TagQuotaValue::fromValue(kv.value);
-						self->trackedTags[tag].setQuota(quota);
+						self->tagStatistics[tag].setQuota(quota);
+						tagsWithQuota.insert(tag);
 					}
-
+					self->removeUnseenQuotas(tagsWithQuota);
 					++self->throttledTagChangeId;
-					// FIXME: Should wait on watch instead
-					// wait(tr.watch(tagThrottleSignalKey));
 					wait(delay(5.0));
 					TraceEvent("GlobalTagThrottler_ChangeSignaled");
 					CODE_PROBE(true, "Global tag throttler detected quota changes");
 					break;
```

Comment:
Are you planning to periodically check the throttle tag change every 5 seconds rather than wait on watch?
If it's the case, I think the `CODE_PROBE` and `TraceEvent("GlobalTagThrottler_ChangeSignaled");` are not accurate.

Reply:
You're right, the code probe originated when we were relying on a watch here, but given that we're now polling every 5 seconds, this is unnecessary. I'll remove it.